### PR TITLE
Introduce number type `RationalRoot` to represent Wigner symbols exactly

### PR DESCRIPTION
**Pull request by w-vdh**

### CODE REVIEW

**Comment by Jutho** [comment]
Looks like a great start, though I am not too fond of the name `AsqrtB` (maybe `RationalRoot` or something similar?) 

I do think the current implementation will still lead to some errors when trying to use this type as yet another `Number` subtype, e.g. adding two `AsqrtB` instances will probably yield a `MethodError`?

Also, some tests would be great ! :-)

File: src/rationalroot.jl [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+export RationalRoot
+
+const IntorRational = Union{Integer,Rational}
+
+"""
+    RationalRoot{T<:Rational} <: Real
+
+Signed square root of a rational number
+"""
+struct RationalRoot{T<:Rational} <: Real
```
Comment (garrison) [inline src/rationalroot.jl:12]: In theory this is the perfect use case for inheriting from `AbstractIrrational`.  May also be worth reviewing discussion at https://github.com/JuliaLang/julia/issues/26550

**Comment by Jutho** [comment]
I did actually prefer your previous implementation, where you keep both the rational outside the square root and the part inside the square root. The drop in test speed could be related to this, as well as to the fact that, if you rely on Base to convert a `Rational{BigInt}` to `Float64`, it goes via `BigFloat`, which is technically a bit safer, but not necessary in most cases. That's why I first convert numerator and denominator to `Float64` and then divide those. I would think that loss of precision is minimal, but I am not too experienced with these matters. It certainly turned out to be faster.

**Comment by w-vdh** [comment]
> I did actually prefer your previous implementation, where you keep both the rational outside the square root and the part inside the square root. 

The problem with that representation is that it is not unique. For example, 
2 * 1/sqrt(3) = 2/3 * sqrt(3)

**Comment by Jutho** [comment]
So I guess it needs to become rational * square root of integer to be unique?

**Comment by garrison** [comment]
> So I guess it needs to become rational * square root of integer to be unique?

and that integer should (obviously) not contain any squares as factors.

**Comment by Jutho** [comment]
I was just testing such an implementation, but it seems that that, indeed, checking that this integer does not contain square prime factors makes the whole thing tremendously slow, as it requires to compute prime factorisations of large numbers just to create the type

**Comment by garrison** [comment]
Rather than factorize, why not run Euclid's algorithm repeatedly on all squares up to n?  My intuition suggests it may be faster than factorization, but i suspect it is still too slow ...

**Comment by garrison** [comment]
then again, is uniqueness essential?

**Comment by Jutho** [comment]
That's something I am also wondering; but maybe the current solution by @w-vdh is just the best way out.

**Comment by Jutho** [comment]
I've started a separate repository here: https://github.com/Jutho/RationalRoots.jl

I mimicked much of the design of [HalfIntegers.jl](https://github.com/sostock/HalfIntegers.jl) (ping @sostock) to make `RationalRoot` into a as complete as possible `Number` type. First order of business is adding tests :-).

**Comment by w-vdh** [comment]
> I was just testing such an implementation, but it seems that that, indeed, checking that this integer does not contain square prime factors makes the whole thing tremendously slow, as it requires to compute prime factorisations of large numbers just to create the type

I encounter the same problem: I let Base.show call for a prime factorization of numerator and denominator of the RationalRoot.

But since WignerSymbols internally uses prime factor representations of the xJ coefficients it should suffice to make these visible to the outside, i.e. to return a type that stores a rational root as a prime factorization. Then Base.show would only have to multiply out the integers (which is currently done inside the wignerxJ functions). This type could be implemented as a RationalRoot, although that would require the representation of a Rational as a prime factorization (allowing both positive and negative powers), and I don't know if this is currently possible in Julia.

**Comment by Jutho** [comment]
WignerSymbols uses the prime factor representation in intermediate calculations, but has to convert it to `BigInt` halfway through the computation in order to add the different fractions together. So the end result is no longer in prime factorised form.

**Comment by Jutho** [comment]
Anyway, RationalRoots.jl is finished, and as soon as it is registered, I have the necessary changes to WignerSymbols.jl ready and will push them. Runtimes are not significantly affected (for better or worse), but the default now is to return the result as a `RationalRoot` of the form in this PR, but then reimplemented along the lines of HalfIntegers.jl in the package RationalRoots.jl

**Comment by Jutho** [comment]
[Latest commit](https://github.com/Jutho/WignerSymbols.jl/commit/8ee830e1737ac12bd16e96ca259d004e9eeed930) adds `RationalRoot` representation as default. Any comments are appreciated. I will likely tag a new version tomorrow or so. Thanks for starting this @w-vdh and my apologies for hijacking your work; you will be mentioned in the release notes.

**Comment by w-vdh** [comment]
No worries.

**Comment by Jutho** [comment]
Thanks again for this; apparently I never closed this PR

**Pull request closed without merging**